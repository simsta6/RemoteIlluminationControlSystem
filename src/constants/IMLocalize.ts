import i18n, { LanguageDetectorAsyncModule } from "i18next";
import { initReactI18next } from "react-i18next";
import AsyncStorage from "@react-native-async-storage/async-storage";
import * as RNLocalize from "react-native-localize";
import { LANGUAGES } from "./languages";
import en from "../translations/en.json";
import lt from "../translations/lt.json";


const LANG_CODES = Object.keys(LANGUAGES.map(lang => lang.code));

const LANGUAGE_DETECTOR: LanguageDetectorAsyncModule = {
    type: "languageDetector",
    async: true,
    detect: (callback) => {
        AsyncStorage.getItem("user-language", (err, language) => {
            if (err || !language) {
                if (err) {
                    console.log("Error fetching Languages from asyncstorage ", err);
                } else {
                    console.log("No language is set, choosing English as fallback");
                }
                const findBestAvailableLanguage = RNLocalize.findBestAvailableLanguage(LANG_CODES);
  
                callback(findBestAvailableLanguage?.languageTag || "en");
                return;
            }
            callback(language);
        });
    },
    init: () => { 
        // later
    },
    cacheUserLanguage: (language: string) => {
        AsyncStorage.setItem("user-language", language);
    }
};

const resources = {
    en,
    lt,
} as const;

i18n
    .use(LANGUAGE_DETECTOR)
    .use(initReactI18next)
    .init({
        compatibilityJSON: "v3",
        lng: "en",
        react: {
            useSuspense: false
        },
        interpolation: {
            escapeValue: false
        },
        resources,
    });