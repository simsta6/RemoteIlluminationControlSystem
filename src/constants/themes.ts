import { Theme } from "@react-navigation/native";

export const KEY_FOR_STORAGE = "defaultScheme";

interface Colors {
    primary: string;
    background: string;
    card: string;
    text: string;
    disabledText: string;
    icon: string;
    border: string;
    notification: string;
    modal: string;
    button: string;
    disabledButton: string;
}

export const AppDarkColors: Colors = {
    primary: "rgba(133, 170, 164, 1)",
    background: "rgba(48, 48, 48, 1)",
    card: "rgba(33, 33, 33, 1)",
    text: "rgba(255, 228, 234, 1)", 
    disabledText: "rgba(186, 168, 172, 1)", 
    icon: "rgba(255, 228, 234, 1)", 
    border: "rgba(51, 70, 69, 1)",
    notification: "rgba(76, 96, 68, 1)",
    modal: "rgba(66, 66, 66, 1)",
    button: "rgba(33, 33, 33, 1)",
    disabledButton: "rgba(74, 74, 74, 1)",
};

export const AppLightColors: Colors = {
    primary: "rgba(122, 85, 91, 1)",
    background: "rgba(255, 235, 236, 1)",
    card: "rgba(229, 208, 237, 1)",
    text: "rgba(122, 85, 91, 1)", 
    disabledText: "rgba(87, 74, 76, 1)", 
    icon: "rgba(122, 85, 91, 1)", 
    border: "rgba(204, 185, 186, 1)",
    notification: "rgba(179, 159, 187, 1)",
    modal: "rgba(255, 247, 248, 1)",
    button: "rgba(229, 208, 237, 1)",
    disabledButton: "rgba(171, 163, 173, 1)",
};

export const AppDarkTheme: Theme = {
    dark: true,
    colors: AppDarkColors,
};

export const AppLightTheme: Theme = {
    dark: false,
    colors: AppLightColors
};