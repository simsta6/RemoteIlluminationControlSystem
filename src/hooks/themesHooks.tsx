import { useDispatch, useSelector } from "react-redux";
import { ThemeActions } from "../state/actions";
import { ActionsUnion, IRootState } from "../state/types";

export const useTheme = () => {
    const { theme } = useSelector((state: IRootState) => state.themeReducer);
    const dispatch: React.Dispatch<ActionsUnion<typeof ThemeActions>> = useDispatch();

    return [
        theme,
        {
            toDarkTheme: () => 
                dispatch(ThemeActions.ChangeToDarkTheme()),
            toLightTheme: () => 
                dispatch(ThemeActions.ChangeToLightTheme()),
        }
    ] as const;
};