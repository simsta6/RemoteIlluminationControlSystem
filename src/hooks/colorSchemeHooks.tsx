import { useTheme } from "@react-navigation/native";
import { AppDarkColors, AppLightColors } from "../constants/themes";

export const useAppColors = () => {
    const { dark } = useTheme();

    return {
        colors: dark ? AppDarkColors : AppLightColors
    };
};
