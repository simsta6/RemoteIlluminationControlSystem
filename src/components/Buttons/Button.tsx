import React from "react";
import { 
    GestureResponderEvent, 
    StyleProp, 
    StyleSheet, 
    Text, 
    TextStyle, 
    TouchableOpacity, 
    ViewStyle
} from "react-native";
import { useAppColors } from "../../hooks/colorSchemeHooks";

interface Props {
    title: string;
    onPress?: ((event: GestureResponderEvent) => void) | undefined;
    disabled?: boolean;
    buttonStyle?: StyleProp<ViewStyle>;
    textStyle?: StyleProp<TextStyle>;
}

export const Button = (props: Props) => {
    const { title, onPress, disabled, buttonStyle, textStyle } = props;
    const { colors } = useAppColors();

    return (
        <TouchableOpacity 
            testID={"ButtonTouchableOpacity"}
            style={[{...styles.button, backgroundColor: disabled ? colors.disabledButton : colors.button }, buttonStyle]} 
            onPress={onPress}
            disabled={disabled}
        >
            <Text 
                testID={"ButtonText"}
                style={[{...styles.buttonText, color: disabled ? colors.disabledText : colors.text}, textStyle]}
            >
                { title }
            </Text>
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    button: {
        alignItems: "center",
        margin: 4,
        padding: 8,  
        borderRadius: 2,
        marginVertical: 5,
    },
    buttonText: {
        textTransform: "uppercase",
        textAlign: "center",
        fontWeight: "500",
    }
});