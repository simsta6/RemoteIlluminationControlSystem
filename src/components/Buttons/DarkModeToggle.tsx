import React from "react";
import { useAppColors } from "../../hooks/colorSchemeHooks";
import { useTheme } from "../../hooks/themesHooks";
import { Switch } from "react-native-switch";
import MoonIcon from "../../assets/icons/MoonIcon";
import SunIcon from "../../assets/icons/SunIcon";
import { StyleSheet, Text, View } from "react-native";
import { useTranslation } from "react-i18next";

export const DarkModeToggle = () => {
    const { t } = useTranslation();
    const { colors } = useAppColors();
    const [scheme, actions] = useTheme();
    const [isEnabled, setIsEnabled] = React.useState(scheme === "dark");

    const toggleSwitch = () => {
        setIsEnabled(previousState => !previousState);
    };

    React.useEffect(() => {
        isEnabled ? actions.toDarkTheme() : actions.toLightTheme();
    }, [isEnabled]);

    return (            
        <>
            <Text style={{...styles.title, color: colors.text}}>{t("DarkModeToggle:darkModeToggle")}</Text>
            <View style={styles.container} testID={"Container"}>
                <Switch
                    renderInsideCircle={() => scheme === "dark" ? 
                        <MoonIcon height={25} width={25} color={colors.icon} /> : 
                        <SunIcon height={25} width={25} color={colors.icon} />
                    }
                    testID={"Switch"}
                    backgroundActive={colors.card}
                    backgroundInactive={colors.card}
                    circleActiveColor={colors.background}
                    circleInActiveColor={colors.background}
                    activeText={""}
                    inActiveText={""}
                    onValueChange={toggleSwitch}
                    value={isEnabled}
                    circleSize={30}
                    circleBorderActiveColor={colors.border}
                    circleBorderInactiveColor={colors.border}
                />
            </View>
        </>
    );
};

const styles = StyleSheet.create({
    container: {
        paddingTop: 10,
    },
    title: {
        fontSize: 20,
        fontWeight: "500"
    },
});