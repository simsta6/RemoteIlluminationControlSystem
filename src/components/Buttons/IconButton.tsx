import React from "react";
import {
    GestureResponderEvent,
    StyleProp, StyleSheet, Text, TextStyle, TouchableOpacity,
    View,
    ViewStyle
} from "react-native";
import { Colors } from "../../constants/themes";
import { useAppColors } from "../../hooks/colorSchemeHooks";

const getText = (
    title: string, 
    disabled: boolean | undefined, 
    colors: Colors, 
    textStyle: StyleProp<TextStyle> | undefined
) => {
    return (
        <Text 
            style={[{...styles.buttonText, color: disabled ? colors.disabledText : colors.text}, textStyle]}
        >
            { title }
        </Text>
    );
}; 

interface Props {
    onPress?: ((event: GestureResponderEvent) => void) | undefined;
    buttonStyle?: StyleProp<ViewStyle>;
    Icon: () => JSX.Element;
    title?: string;
    textStyle?: StyleProp<TextStyle>;
    disabled?: boolean;
}

export const IconButton = (props: Props) => {
    const { onPress, buttonStyle, Icon, title, disabled, textStyle } = props;
    const { colors } = useAppColors();

    return (
        <TouchableOpacity 
            onPress={onPress}
            style={buttonStyle}
        >
            <View style={styles.row}>
                {title && getText(title, disabled, colors, textStyle)}
                <Icon/>
            </View>
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    row: {
        flexDirection: "row",
        alignItems: "center"
    },
    buttonText: {
        textAlign: "center",
        fontWeight: "500",
    }
});