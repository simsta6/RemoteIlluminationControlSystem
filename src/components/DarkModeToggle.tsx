import React from "react";
import { Switch } from "react-native";
import { useTheme } from "../hooks/themesHooks";

export const DarModeToggle = () => {
    const [scheme, actions] = useTheme();
    const [isEnabled, setIsEnabled] = React.useState(scheme === "dark");

    const toggleSwitch = () => {
        setIsEnabled(previousState => !previousState);
    };

    React.useEffect(() => {
        isEnabled ? actions.toDarkTheme() : actions.toLightTheme();
    }, [isEnabled]);

    return (            
        <Switch
            trackColor={{ false: "light", true: "dark" }}
            thumbColor={scheme === "dark" ? "light" : "dark"}
            onValueChange={toggleSwitch}
            value={isEnabled}
        />
    );
};