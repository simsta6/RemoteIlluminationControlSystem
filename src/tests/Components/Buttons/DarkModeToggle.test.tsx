import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { DarkModeToggle } from "../../../components/Buttons/DarkModeToggle";
import { Colors } from "../../../constants/themes";
import * as colorSchemeHooks from "../../../hooks/colorSchemeHooks";
import * as themesHooks from "../../../hooks/themesHooks";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import enzyme from "enzyme";

describe("<DarkModeToggle/>", () => {
    enzyme.configure({ adapter: new Adapter() });
    const colors = {
        icon: "orange",
        text: "purple",
        card: "white",
        background: "red",
        border: "black",
    };

    let scheme: ("dark" | "light") = "dark";

    beforeAll(() => {
        jest.mock("react-i18next", () => ({
            useTranslation: () => {
                return {
                    t: () => "ThemeToggle",
                };
            },
        }));
        jest.spyOn(colorSchemeHooks, "useAppColors").mockImplementation(() => ({ colors: colors as Colors }));

        jest.spyOn(themesHooks, "useTheme").mockImplementation(() => [scheme, {
            toDarkTheme: () => {
                scheme = "dark";
            },
            toLightTheme: () => {
                scheme = "light";
            },
        }]);
    });

    test("Default DarkModeToggle should match snapshot", () => {

        const { toJSON } = render(
            <DarkModeToggle />
        );

        expect(toJSON()).toMatchSnapshot();
    });

    test("should render <MoonIcon/> when dark scheme is selected", () => {

        scheme = "dark";

        const { getByTestId } = render(
            <DarkModeToggle />
        );

        expect(getByTestId("MoonIcon")).toBeTruthy;
    });

    test("should render <SunIcon/> when light scheme is selected", () => {
        
        scheme = "light";

        const { getByTestId } = render(
            <DarkModeToggle />
        );

        expect(getByTestId("SunIcon")).toBeTruthy;
    });

    test("should change theme when toggle is pressed", () => {
        scheme = "dark";
        const { getByTestId } = render(
            <DarkModeToggle />
        );

        fireEvent.press(getByTestId("Switch"));
        expect(scheme).toBe("light");
    });
});