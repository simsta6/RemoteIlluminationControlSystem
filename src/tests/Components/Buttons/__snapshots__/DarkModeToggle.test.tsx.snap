// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DarkModeToggle/> Default DarkModeToggle should match snapshot 1`] = `
Array [
  <Text
    style={
      Object {
        "color": "purple",
        "fontSize": 20,
        "fontWeight": "500",
      }
    }
  >
    DarkModeToggle:darkModeToggle
  </Text>,
  <View
    style={
      Object {
        "paddingTop": 10,
      }
    }
    testID="Container"
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 1)",
          "borderRadius": 30,
          "height": 30,
          "width": 60,
        }
      }
      testID="Switch"
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "left": 15,
            "width": 60,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "white",
              },
              Object {
                "paddingRight": 5,
              },
              undefined,
            ]
          }
        >
          
        </Text>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(255, 0, 0, 1)",
              "borderColor": "rgba(0, 0, 0, 1)",
              "borderRadius": 15,
              "borderWidth": 1,
              "height": 30,
              "justifyContent": "center",
              "width": 30,
            }
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={25}
            bbWidth={25}
            focusable={false}
            height={25}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 25,
                  "width": 25,
                },
              ]
            }
            testID="MoonIcon"
            vbHeight={512}
            vbWidth={512}
            width={25}
            x="0"
            y="0"
          >
            <RNSVGGroup>
              <RNSVGPath
                d="M275.4 500.7c-135 0-244.7-109.8-244.7-244.7 0-134.9 109.8-244.7 244.7-244.7 8.2 0 16.4.4 24.6 1.2 7.2.7 13.5 5.2 16.5 11.7s2.4 14.2-1.6 20.2c-23 33.8-35.2 73.3-35.2 114.2 0 105 78.7 192.2 183.2 202.6 7.2.7 13.5 5.2 16.5 11.7 3.1 6.5 2.4 14.2-1.6 20.2-45.8 67.4-121.4 107.6-202.4 107.6zm-12.5-448C156.4 59.2 71.7 147.9 71.7 256c0 112.3 91.4 203.7 203.7 203.7 56.4 0 109.6-23.4 147.8-63.7-46.2-11.7-88.1-36.8-120.8-72.6-41.1-45.2-63.8-103.6-63.8-164.6.1-37.1 8.4-73.2 24.3-106.1z"
                fill={4294944000}
                propList={
                  Array [
                    "fill",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </View>,
]
`;
